File: language/runtime/ceylon/language/List.java
package ceylon.language;

import com.redhat.ceylon.compiler.java.metadata.Annotation;
import com.redhat.ceylon.compiler.java.metadata.Annotations;
import com.redhat.ceylon.compiler.java.metadata.Ceylon;
import com.redhat.ceylon.compiler.java.metadata.Name;
import com.redhat.ceylon.compiler.java.metadata.SatisfiedTypes;
import com.redhat.ceylon.compiler.java.metadata.TypeInfo;
import com.redhat.ceylon.compiler.java.metadata.TypeParameter;
import com.redhat.ceylon.compiler.java.metadata.TypeParameters;
import com.redhat.ceylon.compiler.java.metadata.Variance;

@Ceylon(major = 4)
@TypeParameters(@TypeParameter(value = "Element", variance = Variance.OUT))
@SatisfiedTypes({"ceylon.language::Collection<Element>",
                 "ceylon.language::Correspondence<ceylon.language::Integer,Element>",
                 "ceylon.language::Ranged<ceylon.language::Integer,ceylon.language::List<Element>>",
                 "ceylon.language::Cloneable<ceylon.language::List<Element>>"})
public interface List<Element>
        extends Collection<Element>,
                Correspondence<Integer,Element>,
                Ranged<Integer, List<? extends Element>> {

    @Annotations(@Annotation("formal"))
    @TypeInfo("ceylon.language::Null|ceylon.language::Integer")
    public Integer getLastIndex();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @Override
    public long getSize();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @Override
    public boolean defines(@Name("index") Integer key);

    @Annotations({@Annotation("actual"), @Annotation("formal")})
    @TypeInfo("ceylon.language::Null|Element")
    @Override
    public Element item(@Name("index") Integer key);

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("ceylon.language::Iterator<Element>")
    @Override
    public Iterator<? extends Element> getIterator();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @Override
    public boolean equals(@Name("that") @TypeInfo("ceylon.language::Object")
    java.lang.Object that);

    @Annotations({@Annotation("formal")})
    @TypeInfo("ceylon.language::List<Element>")
    public List<? extends Element> getReversed();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @Override
    public int hashCode();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("Element|ceylon.language::Null")
    public Element findLast(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language.Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("Element|ceylon.language::Null")
    public Element getFirst();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("Element|ceylon.language::Null")
    public Element getLast();
    
    @SuppressWarnings("rawtypes")
    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeParameters(@TypeParameter("Other"))
    @TypeInfo("ceylon.language::Sequence<Element|Other>")
    public <Other> Sequence withLeading(@Name("element")
            @TypeInfo("Other")
            Other element);

    @SuppressWarnings("rawtypes")
    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeParameters(@TypeParameter("Other"))
    @TypeInfo("ceylon.language::Sequence<Element|Other>")
    public <Other> Sequence withTrailing(@Name("element")
            @TypeInfo("Other")
            Other element);

}

File: language/runtime-js/replaced/list.js
function List(wat) {
    Collection(wat);
    Correspondence(wat);
    Ranged(wat);
    return wat;
}
function $init$List() {
    if (List.$$===undefined) {
        initTypeProto(List, 'ceylon.language::List', $init$Collection(), $init$Correspondence(),
            $init$Ranged());
    }
var List$proto = List.$$.prototype;
List$proto.getSize = function() {
    var li = this.getLastIndex();
    return li === null ? 0 : li.getSuccessor();
}
List$proto.defines = function(idx) {
    var li = this.getLastIndex();
    if (li === null) li = -1;
    return li.compare(idx) !== smaller;
}
List$proto.equals = function(other) {
    if (isOfType(other, {t:List}) && other.getSize().equals(this.getSize())) {
        for (var i = 0; i < this.getSize(); i++) {
            var mine = this.get(i);
            var theirs = other.get(i);
            if (((mine === null) && theirs) || !(mine && mine.equals(theirs))) {
                return false;
            }
        }
        return true;
    }
    return false;
}
List$proto.getHash = function() {
    var hc=1;
    var iter=this.getIterator();
    var e; while ((e = iter.next()) != getFinished()) {
        hc*=31;
        if (e !== null) {
            hc += e.getHash();
        }
    }
    return hc;
}
List$proto.findLast = function(select) {
    var li = this.getLastIndex();
    if (li !== null) {
        while (li>=0) {
            var e = this.get(li);
            if (e !== null && select(e)) {
                return e;
            }
            li = li.getPredecessor();
        }
    }
    return null;
}
List$proto.withLeading = function(other, $$$mptypes) {
    var sb = SequenceBuilder({Element:{t:'u',l:[this.$$targs$$.Element, $$$mptypes.Other]}});
    sb.append(other);
    sb.appendAll(this);
    return sb.getSequence();
}
List$proto.withTrailing = function(other, $$$mptypes) {
    var sb = SequenceBuilder({Element:{t:'u',l:[this.$$targs$$.Element, $$$mptypes.Other]}});
    sb.appendAll(this);
    sb.append(other);
    return sb.getSequence();
}
exports.List=List;

List$proto.getIterator = function() {
    return ListIterator(this);
}
    return List;
}
$init$List();
function ListIterator(list) {
    var that = new ListIterator.$$;
    Iterator(that);
    that.list=list;
    that.$$targs$$=list.$$targs$$;
    that.index=0;
    that.lastIndex=list.getLastIndex();
    if (that.lastIndex === null) {
        that.lastIndex = -1;
    } else {
        that.lastIndex = that.lastIndex;
    }
    return that;
}
initTypeProto(ListIterator, 'ceylon.language::ListIterator', $init$Iterator());
ListIterator.$$.prototype.next = function() {
    if (this.index <= this.lastIndex) {
        return this.list.get(this.index++);
    }
    return getFinished();
}
